File: src/runners/routing.ts
/**
 * Runner routing (§16, extended with DeepSeek BYOK).
 *
 * | Signal | Where it runs |
 * |---|---|
 * | Control plane `busy` | prefer `local` if a local runner is healthy |
 * | `runnerPreference = local` | local runner only |
 * | model `deepseek-*` (workers) | DeepSeek cloud (BYOK `DEEPSEEK_API_KEY`) |
 * | otherwise (planners) | Claude cloud (`ANTHROPIC_API_KEY`) |
 * | Per-goal override | wins over default |
 *
 * When a cloud backend is requested but its key is absent, we fall back to
 * the simulated engine and LABEL the session accordingly (demo/test mode) —
 * the recorded `runner` field keeps the requested kind.
 */
import type { Runner } from "./types.js";
import type { GoalRunnerPreference, RunnerPreference, RunnerKind } from "../domain/types.js";
import { CloudClaudeRunner } from "./cloud.js";
import { LocalVmRunner } from "./local.js";
import { SimulatedRunner } from "./simulated.js";
import { DeepseekRunner } from "./deepseek.js";
import { isDeepseekModel } from "./models.js";

export interface RunnerChoice {
  /** Recorded runner kind: cloud | local | deepseek. */
  kind: RunnerKind;
  /** The backend that will actually execute. */
  runner: Runner;
  /** Human-readable note (e.g. simulated fallback). */
  note: string;
}

export function chooseRunner(opts: {
  agentPreference: RunnerPreference;
  goalPreference?: GoalRunnerPreference | null;
  cloudBusy?: boolean;
  model: string;
  anthropicApiKey?: string;
  deepseekApiKey?: string;
  deepseekBaseUrl?: string;
  localMode?: "inprocess" | "worker";
}): RunnerChoice {
  const {
    agentPreference,
    goalPreference,
    cloudBusy = false,
    model,
    anthropicApiKey,
    deepseekApiKey,
    deepseekBaseUrl,
    localMode = "inprocess",
  } = opts;

  // Per-goal override wins (§16).
  const effective: RunnerPreference | GoalRunnerPreference =
    goalPreference === "local" || goalPreference === "cloud"
      ? goalPreference
      : goalPreference === "auto"
        ? agentPreference === "inherit" ? "cloud" : agentPreference
        : agentPreference === "inherit"
          ? "cloud"
          : agentPreference;

  if (effective === "local") {
    return {
      kind: "local",
      runner: new LocalVmRunner(localMode),
      note: "routed local (preference)",
    };
  }

  // Cloud routing is model-driven.
  if (isDeepseekModel(model)) {
    if (deepseekApiKey) {
      return {
        kind: "deepseek",
        runner: new DeepseekRunner({ apiKey: deepseekApiKey, baseUrl: deepseekBaseUrl }),
        note: "routed deepseek (BYOK)",
      };
    }
    return {
      kind: "deepseek",
      runner: new SimulatedRunner(),
      note: `deepseek model "${model}" but no DEEPSEEK_API_KEY — simulated stand-in (labeled)`,
    };
  }

  // Claude / default.
  if (anthropicApiKey) {
    return {
      kind: "cloud",
      runner: new CloudClaudeRunner(),
      note: cloudBusy ? "routed cloud (busy policy prefers local; cloud used)" : "routed cloud",
    };
  }
  return {
    kind: "cloud",
    runner: new SimulatedRunner(),
    note: "cloud requested but no ANTHROPIC_API_KEY — simulated stand-in (labeled)",
  };
}

File: agentos/runners/routing.py
"""
Runner routing. Port of src/runners/routing.ts.
Model-driven: deepseek-* → DeepSeekRunner, else → CloudClaudeRunner.
Falls back to SimulatedRunner when the required API key is absent.
"""
from __future__ import annotations

from agentos.runners.models import is_deepseek_model


def choose_runner(
    *,
    agent_preference: str,
    goal_preference: str | None = None,
    cloud_busy: bool = False,
    model: str,
    anthropic_api_key: str | None = None,
    deepseek_api_key: str | None = None,
    deepseek_base_url: str = "https://api.deepseek.com",
    local_mode: str = "inprocess",
) -> dict:
    """
    Returns {"kind": RunnerKind, "runner": Runner, "note": str}.
    """
    from agentos.runners.cloud import CloudClaudeRunner
    from agentos.runners.deepseek import DeepSeekRunner
    from agentos.runners.simulated import SimulatedRunner

    # Effective preference: goal overrides agent; "auto"/"inherit" → cloud
    effective = goal_preference if goal_preference in ("local", "cloud") else (
        agent_preference if agent_preference != "inherit" else "cloud"
    )

    if effective == "local":
        # ponytail: local VM runner not yet implemented — simulated stand-in
        return {
            "kind": "local",
            "runner": SimulatedRunner(),
            "note": "local runner requested — simulated stand-in (not yet implemented)",
        }

    if is_deepseek_model(model):
        if deepseek_api_key:
            return {
                "kind": "deepseek",
                "runner": DeepSeekRunner(api_key=deepseek_api_key, base_url=deepseek_base_url),
                "note": "routed deepseek (BYOK)",
            }
        return {
            "kind": "deepseek",
            "runner": SimulatedRunner(),
            "note": f'deepseek model "{model}" but no DEEPSEEK_API_KEY — simulated stand-in (labeled)',
        }

    if anthropic_api_key:
        return {
            "kind": "cloud",
            "runner": CloudClaudeRunner(api_key=anthropic_api_key),
            "note": "routed cloud (Claude)",
        }
    return {
        "kind": "cloud",
        "runner": SimulatedRunner(),
        "note": "cloud requested but no ANTHROPIC_API_KEY — simulated stand-in (labeled)",
    }
